framework/auth/views.py

```diff
@@ -251,6 +287,62 @@ def confirm_email_get(token, auth=None, **kwargs):
     ))
 
 
+@collect_auth
+def verified_email_remove(auth=None):
+    """Called at login if user cancels their merge or email add.
+    methods: DELETE
+    """
+    user = auth.user
+    confirmed_email = request.get_json()
+    email_verifications = deepcopy(user.email_verifications)
+    for token in user.email_verifications:
+        if token == confirmed_email['token']:
+            email_verifications.pop(token)
+    user.email_verifications = email_verifications
+    user.save()
+    return {
+        'status': 'success',
+        'removed_email': confirmed_email
+    }, 200
+
+
+@collect_auth
+def verified_email_add(auth=None):
+    """Called at login if user confirms their merge or email add.
+    methods: PUT
+    """
+    user = auth.user
+    token = request.json.get('token')
+
+    if user is None:
+        raise HTTPError(http.NOT_FOUND)
+
+    try:
+        user.confirm_email(token, merge=True)
+    except exceptions.EmailConfirmTokenError as e:
+        raise HTTPError(http.BAD_REQUEST, data={
+            'message_short': e.message_short,
+            'message_long': e.message_long
+        })
+
+    if token:
+        user.save()
+        return {
+            'status': 'success',
+        }, 200
+
+    # Redirect to CAS and authenticate the user with a verification key.
+    user.verification_key = security.random_string(20)
+    user.save()
+
+    return redirect(cas.get_login_url(
```

Comment:
Does this redirect actually do anything? It appears incorrect because this is a JSON endpoint.